x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/regex/AbstractCompositeExpression.java

```diff
@@ -0,0 +1,287 @@
+     * @param transformer transformer to use
+     * @return result of transforming components
+     */
+    protected <J> Set<J> transformComponents(Expression.Transformer<T, J> transformer) {
+        Set<J> builder = new HashSet<>();
+        for (Expression<T> component : components) {
+            builder.add(component.transform(transformer));
+        }
+        return Collections.unmodifiableSet(builder);
+    }
+
+    /**
+     * Build an expression of this type with a list of components. Used in
+     * simplification.
+     */
+    protected abstract AbstractCompositeExpression<T> newFrom(Set<Expression<T>> components);
+
+    /**
+     * Does this component of this expression affect the outcome of the overall
+     * expression? For example the TRUE in (TRUE AND foo) doesn't effect the
+     * expression and can be simplified away.
+     */
+    protected abstract boolean doesNotAffectOutcome(Expression<T> expression);
+
+    /**
+     * Does this component force the expression to a certain value? For example
+     * the FALSE in (FALSE and foo) forces the whole expression to FALSE.
+     */
+    protected abstract Expression<T> componentForcesOutcome(Expression<T> expression);
+
+    protected abstract String toStringJoiner();
+
+    @Override
+    public boolean alwaysFalse() {
+        return false;
+    }
+
+    @Override
+    public boolean alwaysTrue() {
+        return false;
+    }
+
+    @Override
+    public Expression<T> simplify() {
+        if (simplified) {
+            return this;
+        }
+        Iterator<Expression<T>> componentsItr = components.iterator();
+        List<Expression<T>> newComponentsBuilder = null;
+        boolean changed = false;
+        while (componentsItr.hasNext()) {
+            Expression<T> expression = componentsItr.next();
+            Expression<T> simplified = expression.simplify();
+            changed |= expression != simplified;
+            if (doesNotAffectOutcome(simplified)) {
+                changed |= true;
+                continue;
+            }
+            Expression<T> forcedOutcome = componentForcesOutcome(simplified);
+            if (forcedOutcome != null) {
+                return forcedOutcome;
+            }
+            if (newComponentsBuilder == null) {
+                newComponentsBuilder = new ArrayList<>(components.size());
+            }
+            if (simplified.getClass() == getClass()) {
+                changed |= true;
+                newComponentsBuilder.addAll(((AbstractCompositeExpression<T>) simplified).components);
+                continue;
+            }
+            newComponentsBuilder.add(simplified);
+        }
+        if (newComponentsBuilder == null) {
+            // Nothing left in the expression!
+            return True.instance();
+        }
+        switch (newComponentsBuilder.size()) {
+        case 0:
+            return True.instance();
+        case 1:
+            return newComponentsBuilder.get(0);
+        default:
```

Comment:
This'll piss off someone's IDE. Probably best to leave a comment about how the fall through is intentional.